pkg/kube/krt/join.go

```diff
@@ -93,8 +122,16 @@ type joinIndexer struct {
 // nolint: unused // (not true)
 func (j joinIndexer) Lookup(key string) []any {
 	var res []any
+	first := true
 	for _, i := range j.indexers {
-		res = append(res, i.Lookup(key)...)
+		l := i.Lookup(key)
+		if len(l) > 0 && first {
+			// Optimization: re-use the first returned slice
+			res = l
+			first = false
+		} else {
+			res = append(res, l...)
+		}
 	}
 	return res
```

Comment:
Can joinIndexer benefit from a similar optimization to indicate that keys are unique so we can just return as soon as we get a match if that boolean is true?

Reply:
Maybe but the same config could not apply - we would need a new one tied to the index creation, which would be a pretty sweeping change.

That also requires the _user_ of a collection to know its a join under the hood which is a bit of a layering abstraction

Reply:
Maybe I misunderstand but it seems like a `joinIndexer` produced from a `join` which has `uncheckedOverlap` should not have overlapping keys and could return on the first match (which should be the only match). It would have the same basic rules, if you don't know what to do leave it as false to get safe behavior which might not be fully optimal and if you set it to true where keys are not unique then Lookup produces undefined behavior.

Reply:
the keys in the index are unrelated to the keys in the underlying collection. If anything it actually maybe needs to dedupe check when we *don't* set the unchecked field now that I think about it..

Reply:
for example you may index on namespace and have the key be "istio-system" which will definitely return multiple values

Reply:
I see now. This would need to be managed/understood when defining the extract function (which shouldn't have to know/care if the implementing type is joined)... yeah that's going to be pretty unergonomic at best I think.

Thanks for the explanation.